study1/week4/flood-fill.java

```diff
@@ -0,0 +1,30 @@
+// https://leetcode.com/problems/flood-fill/
+// Time Complexity: O(n*m), n: image.length, m: image[0].length
+// Space Complexity: O(1)
+// using recursive methods
+class Solution {
+    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
+        if (image[sr][sc] == color) {
+            return image;
+        }
+
+        int pixelColor = image[sr][sc]; // original color
+
+        image[sr][sc] = color; // change the color of the pixel
+
+        int[][] dir = new int[][]{{0, 1},{0, -1},{1, 0},{-1, 0}}; // up, down, right, left
+        for (int i = 0; i < dir.length; ++i) {
+            int nr = sr + dir[i][0];
+            int nc = sc + dir[i][1];
+
+            if (nr < 0 || nr >= image.length || nc < 0 || nc >= image[0].length) {
```

Comment:
This simplifies the code, but it's better to unroll the loop and do it one by one. For instance, you don't have to compare nc when you're going up or down, and you don't need to compare nr when you're going left and right.

So the loop in line 16 should be removed and become : 

```java
if (sr < image.length - 1 && image[sr + 1][sc] == pixelColor) {
   floodFill(image, sr + 1, sc, color)  
}
if (sr >0 && image[sr - 1][sc] == pixelColor) {
   floodFill(image, sr - 1, sc, color)  
}
if (sc < image.length - 1 && image[sr][sc + 1] == pixelColor) {
   floodFill(image, sr, sc + 1, color)  
}
if (sc > 0 && image[sr][sc - 1] == pixelColor) {
   floodFill(image, sr, sc - 1, color)  
}
```